File: Striver/Trees/Q09.java
package Striver.Trees;
import java.util.*;

public class Q09{
    // Preorder, Inorder, Postorder Traversal in a single iteration using a single stack

    public void preInPost(Node root){
        Stack<Pair> st = new Stack<>();
        st.push(new Pair(root, 1));

        List<Integer> pre = new ArrayList<>();
        List<Integer> in = new ArrayList<>();
        List<Integer> post = new ArrayList<>();

        if(root == null) return;

        while(!st.isEmpty()){
            Pair top = st.peek();
            if(top.state == 1){
                pre.add(top.node.data);
                top.state++;
                if(top.node.left != null){
                    st.push(new Pair(top.node.left, 1));
                }
            }else if(top.state == 2){
                in.add(top.node.data);
                top.state++;
                if(top.node.right != null){
                    st.push(new Pair(top.node.right, 1));
                }
            }else{
                post.add(top.node.data);
                st.pop();
            }
        }

        System.out.println("Preorder: " + pre);
        System.out.println("Inorder: " + in);
        System.out.println("Postorder: " + post);
    }

    public static void main(String[] args) {
        Node root = new Node(10);
        root.left = new Node(20);
        root.left.left = new Node(40);
        root.left.right = new Node(50);
        root.right = new Node(30);
        root.right.left = new Node(60);
        root.right.right = new Node(70);
        
        Q09 q = new Q09();
        q.preInPost(root);
    }
}

File: Striver/Trees/Q09.py
# preorder, inorder, postorder traversal of a tree in a single iteration using single stack

from Node import Node

def preInPost(root):
    st = []
    st.append([root, 1])

    if not root:
        return

    pre = []
    ino = []
    post = []

    while len(st) > 0:
        top = st[-1]
        if top[1] == 1:
            pre.append(top[0].data)
            top[1] += 1
            if top[0].left:
                st.append([top[0].left, 1])
        elif top[1] == 2:
            ino.append(top[0].data)
            top[1] += 1
            if top[0].right:
                st.append([top[0].right, 1])
        else:
            post.append(top[0].data)
            st.pop()

    print("Preorder: ", pre)
    print("Inorder: ", ino)
    print("Postorder: ", post)

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.left.left = Node(8)
root.left.left.right = Node(9)
root.left.right = Node(5)
root.right.left = Node(6)
root.right.right = Node(7)

preInPost(root)